File: src/actions/home_assistant/connector/websocket.py
"""Home Assistant WebSocket API action connector."""

import json
import logging
from typing import Optional

import websockets
from pydantic import Field

from actions.base import ActionConfig, ActionConnector
from actions.home_assistant.interface import (
    COLOR_MAP,
    HAAction,
    HADeviceType,
    HomeAssistantInput,
)


class HomeAssistantWebSocketConfig(ActionConfig):
    """
    Configuration for Home Assistant WebSocket connector.

    Parameters
    ----------
    base_url : str
        Home Assistant base URL (e.g. http://homeassistant.local:8123).
    token : str
        Long-lived access token from Home Assistant profile.
    timeout : float
        Connection and command timeout in seconds.
    """

    base_url: str = Field(default="", description="Home Assistant base URL")
    token: str = Field(default="", description="Long-lived access token")
    timeout: float = Field(default=10.0, description="Timeout in seconds")


class HomeAssistantWebSocketConnector(
    ActionConnector[HomeAssistantWebSocketConfig, HomeAssistantInput]
):
    """
    Home Assistant connector using WebSocket API.

    Provides persistent bidirectional connection to Home Assistant,
    enabling real-time command execution with result confirmation.
    """

    def __init__(self, config: HomeAssistantWebSocketConfig):
        """
        Initialize the WebSocket connector.

        Parameters
        ----------
        config : HomeAssistantWebSocketConfig
            Configuration for the connector.
        """
        super().__init__(config)
        base = config.base_url.rstrip("/")
        if base.startswith("https://"):
            self._ws_url = "wss://" + base[len("https://") :] + "/api/websocket"
        elif base.startswith("http://"):
            self._ws_url = "ws://" + base[len("http://") :] + "/api/websocket"
        else:
            self._ws_url = base + "/api/websocket"
        self._token = config.token
        self._timeout = config.timeout
        self._msg_id = 1

        if not config.base_url:
            logging.warning("HomeAssistantWebSocketConnector: base_url not provided")
        if not config.token:
            logging.warning("HomeAssistantWebSocketConnector: token not provided")

    async def _send_command(
        self,
        domain: str,
        service: str,
        entity_id: str,
        service_data: Optional[dict] = None,
    ) -> bool:
        """
        Open a WebSocket connection, authenticate, and send a service command.

        Parameters
        ----------
        domain : str
            Service domain (e.g. 'light', 'switch', 'climate').
        service : str
            Service name (e.g. 'turn_on', 'turn_off').
        entity_id : str
            Target entity ID.
        service_data : Optional[dict]
            Additional service data (brightness, temperature, etc.).

        Returns
        -------
        bool
            True if command succeeded, False otherwise.
        """
        if not self._ws_url or not self._token:
            return False

        try:
            async with websockets.connect(
                self._ws_url, open_timeout=self._timeout
            ) as ws:
                raw = await ws.recv()
                msg = json.loads(raw)
                if msg.get("type") != "auth_required":
                    logging.error(
                        f"HomeAssistantWebSocketConnector: expected auth_required, "
                        f"got {msg.get('type')}"
                    )
                    return False

                await ws.send(json.dumps({"type": "auth", "access_token": self._token}))
                raw = await ws.recv()
                msg = json.loads(raw)
                if msg.get("type") != "auth_ok":
                    logging.error(
                        "HomeAssistantWebSocketConnector: authentication failed"
                    )
                    return False

                command = {
                    "id": self._msg_id,
                    "type": "call_service",
                    "domain": domain,
                    "service": service,
                    "target": {"entity_id": entity_id},
                    "service_data": service_data or {},
                }
                self._msg_id += 1

                await ws.send(json.dumps(command))
                raw = await ws.recv()
                result = json.loads(raw)

                if result.get("success"):
                    return True
                else:
                    error = result.get("error", {})
                    logging.error(
                        f"HomeAssistantWebSocketConnector: command failed - "
                        f"{error.get('code')}: {error.get('message')}"
                    )
                    return False

        except TimeoutError:
            logging.error("HomeAssistantWebSocketConnector: connection timed out")
            return False
        except websockets.exceptions.WebSocketException as e:
            logging.error(f"HomeAssistantWebSocketConnector: WebSocket error - {e}")
            return False
        except Exception as e:
            logging.error(f"HomeAssistantWebSocketConnector: unexpected error - {e}")
            return False

    async def connect(self, output_interface: HomeAssistantInput) -> None:
        """
        Execute a Home Assistant device control action via WebSocket.

        Parameters
        ----------
        output_interface : HomeAssistantInput
            The action request containing device_type, entity_id, action,
            and optional parameters.
        """
        device_type = output_interface.device_type
        entity_id = output_interface.entity_id
        action = output_interface.action

        if device_type == HADeviceType.LIGHT:
            if action == HAAction.TURN_ON:
                await self._send_command("light", "turn_on", entity_id)
            elif action == HAAction.TURN_OFF:
                await self._send_command("light", "turn_off", entity_id)
            elif action == HAAction.SET_BRIGHTNESS:
                brightness = output_interface.brightness or 255
                await self._send_command(
                    "light", "turn_on", entity_id, {"brightness": brightness}
                )
            elif action == HAAction.SET_COLOR:
                color = (output_interface.color or "white").lower()
                hs_color = COLOR_MAP.get(color, COLOR_MAP["white"])
                await self._send_command(
                    "light", "turn_on", entity_id, {"hs_color": hs_color}
                )
            else:
                logging.warning(
                    f"HomeAssistantWebSocketConnector: action '{action.value}' "
                    f"not supported for light"
                )

        elif device_type == HADeviceType.SWITCH:
            if action == HAAction.TURN_ON:
                await self._send_command("switch", "turn_on", entity_id)
            elif action == HAAction.TURN_OFF:
                await self._send_command("switch", "turn_off", entity_id)
            else:
                logging.warning(
                    f"HomeAssistantWebSocketConnector: action '{action.value}' "
                    f"not supported for switch"
                )

        elif device_type == HADeviceType.CLIMATE:
            if action == HAAction.SET_TEMPERATURE:
                temperature = output_interface.temperature or 20.0
                await self._send_command(
                    "climate",
                    "set_temperature",
                    entity_id,
                    {"temperature": temperature},
                )
            elif action == HAAction.TURN_ON:
                await self._send_command("climate", "turn_on", entity_id)
            elif action == HAAction.TURN_OFF:
                await self._send_command("climate", "turn_off", entity_id)
            else:
                logging.warning(
                    f"HomeAssistantWebSocketConnector: action '{action.value}' "
                    f"not supported for climate"
                )

        else:
            logging.warning(
                f"HomeAssistantWebSocketConnector: device_type '{device_type.value}' "
                f"not supported"
            )

File: plugins/actions/home_assistant/websocket.go
package home_assistant

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/gorilla/websocket"
	"go.uber.org/zap"

	"github.com/openmind/om1/internal/actions"
	"github.com/openmind/om1/internal/logger"
)

const (
	defaultWSTimeout     = 10 * time.Second
	defaultWSBrightness  = 255
	defaultWSTemperature = 20.0
)

// WSConfig is the decoded plugin configuration for the WebSocket connector.
type WSConfig struct {
	// BaseURL is the root URL of the Home Assistant instance
	// (e.g. "http://homeassistant.local:8123"). Trailing slashes are ignored.
	BaseURL string `json:"base_url"`

	// Token is the Home Assistant long-lived access token.
	Token string `json:"token"`

	// Timeout is the connection/command timeout in seconds. 0 means use the default.
	Timeout float64 `json:"timeout"`
}

// WSConnector implements actions.Connector using the Home Assistant
// WebSocket API. Each Connect call opens a fresh connection, authenticates,
// sends one call_service command, and closes the connection.
type WSConnector struct {
	log     *zap.Logger
	wsURL   string
	token   string
	timeout time.Duration
	msgID   int
}

func init() {
	actions.Register("home_assistant/websocket", NewWSConnector)
}

// buildWSURL converts an http(s) base URL to a ws(s) Home Assistant
// websocket endpoint.
func buildWSURL(baseURL string) string {
	base := strings.TrimRight(baseURL, "/")
	switch {
	case strings.HasPrefix(base, "https://"):
		return "wss://" + strings.TrimPrefix(base, "https://") + "/api/websocket"
	case strings.HasPrefix(base, "http://"):
		return "ws://" + strings.TrimPrefix(base, "http://") + "/api/websocket"
	default:
		return base + "/api/websocket"
	}
}

// NewWSConnector builds the connector from a decoded config map.
func NewWSConnector(configMap map[string]any) (actions.Connector, error) {
	var cfg WSConfig
	if b, err := json.Marshal(configMap); err == nil {
		_ = json.Unmarshal(b, &cfg)
	}

	log := logger.Get().Named("home_assistant/websocket")

	if cfg.BaseURL == "" {
		log.Warn("base_url not provided")
	}
	if cfg.Token == "" {
		log.Warn("token not provided")
	}

	timeout := defaultWSTimeout
	if cfg.Timeout > 0 {
		timeout = time.Duration(cfg.Timeout * float64(time.Second))
	}

	return &WSConnector{
		log:     log,
		wsURL:   buildWSURL(cfg.BaseURL),
		token:   cfg.Token,
		timeout: timeout,
		msgID:   1,
	}, nil
}

// sendCommand opens a WebSocket connection, authenticates, sends a single
// call_service command, and reports whether it succeeded.
func (c *WSConnector) sendCommand(ctx context.Context, domain, service, entityID string, serviceData map[string]any) bool {
	if c.wsURL == "" || c.token == "" {
		return false
	}

	dialer := websocket.Dialer{HandshakeTimeout: c.timeout}
	conn, _, err := dialer.DialContext(ctx, c.wsURL, nil)
	if err != nil {
		c.log.Error("connection timed out or failed", zap.Error(err))
		return false
	}
	defer func() { _ = conn.Close() }()

	_ = conn.SetReadDeadline(time.Now().Add(c.timeout))

	// Step 1: expect auth_required
	var authRequired map[string]any
	if err := conn.ReadJSON(&authRequired); err != nil {
		c.log.Error("WebSocket error", zap.Error(err))
		return false
	}
	if authRequired["type"] != "auth_required" {
		c.log.Error("expected auth_required", zap.Any("got", authRequired["type"]))
		return false
	}

	// Step 2: send auth
	if err := conn.WriteJSON(map[string]any{
		"type":         "auth",
		"access_token": c.token,
	}); err != nil {
		c.log.Error("WebSocket error", zap.Error(err))
		return false
	}

	var authResult map[string]any
	if err := conn.ReadJSON(&authResult); err != nil {
		c.log.Error("WebSocket error", zap.Error(err))
		return false
	}
	if authResult["type"] != "auth_ok" {
		c.log.Error("authentication failed")
		return false
	}

	// Step 3: send call_service command
	if serviceData == nil {
		serviceData = map[string]any{}
	}
	msgID := c.msgID
	c.msgID++

	command := map[string]any{
		"id":      msgID,
		"type":    "call_service",
		"domain":  domain,
		"service": service,
		"target": map[string]any{
			"entity_id": entityID,
		},
		"service_data": serviceData,
	}
	if err := conn.WriteJSON(command); err != nil {
		c.log.Error("WebSocket error", zap.Error(err))
		return false
	}

	var result map[string]any
	if err := conn.ReadJSON(&result); err != nil {
		c.log.Error("WebSocket error", zap.Error(err))
		return false
	}

	if success, _ := result["success"].(bool); success {
		return true
	}

	errInfo, _ := result["error"].(map[string]any)
	c.log.Error("command failed",
		zap.Any("code", errInfo["code"]),
		zap.Any("message", errInfo["message"]),
	)
	return false
}

// Connect resolves the requested device action and sends the corresponding
// Home Assistant WebSocket command.
func (c *WSConnector) Connect(ctx context.Context, input actions.Input) (actions.Output, error) {
	args, ok := input.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("home_assistant/websocket: unexpected input type %T", input)
	}

	entityID, _ := args["entity_id"].(string)
	deviceType, _ := args["device_type"].(string)
	action, _ := args["action"].(string)
	brightness, _ := args["brightness"].(string)
	temperature, _ := args["temperature"].(string)
	color, _ := args["color"].(string)

	if entityID == "" {
		return nil, nil
	}

	switch deviceType {
	case "light":
		c.connectLight(ctx, entityID, action, brightness, color)
	case "switch":
		c.connectSwitch(ctx, entityID, action)
	case "climate":
		c.connectClimate(ctx, entityID, action, temperature)
	default:
		c.log.Warn("device_type not supported", zap.String("device_type", deviceType))
	}

	return nil, nil
}

func (c *WSConnector) connectLight(ctx context.Context, entityID, action, brightness, color string) {
	switch action {
	case "turn_on":
		c.sendCommand(ctx, "light", "turn_on", entityID, nil)
	case "turn_off":
		c.sendCommand(ctx, "light", "turn_off", entityID, nil)
	case "set_brightness":
		v := parseBrightness(brightness)
		c.sendCommand(ctx, "light", "turn_on", entityID, map[string]any{"brightness": v})
	case "set_color":
		hs := resolveHSColor(color)
		c.sendCommand(ctx, "light", "turn_on", entityID, map[string]any{"hs_color": []int{hs[0], hs[1]}})
	default:
		c.log.Warn("action not supported for light", zap.String("action", action))
	}
}

func (c *WSConnector) connectSwitch(ctx context.Context, entityID, action string) {
	switch action {
	case "turn_on":
		c.sendCommand(ctx, "switch", "turn_on", entityID, nil)
	case "turn_off":
		c.sendCommand(ctx, "switch", "turn_off", entityID, nil)
	default:
		c.log.Warn("action not supported for switch", zap.String("action", action))
	}
}

func (c *WSConnector) connectClimate(ctx context.Context, entityID, action, temperature string) {
	switch action {
	case "turn_on":
		c.sendCommand(ctx, "climate", "turn_on", entityID, nil)
	case "turn_off":
		c.sendCommand(ctx, "climate", "turn_off", entityID, nil)
	case "set_temperature":
		t := parseTemperature(temperature, defaultWSTemperature)
		c.sendCommand(ctx, "climate", "set_temperature", entityID, map[string]any{"temperature": t})
	default:
		c.log.Warn("action not supported for climate", zap.String("action", action))
	}
}

// Tick blocks until ctx is cancelled; this connector is event-driven.
func (c *WSConnector) Tick(ctx context.Context) {
	<-ctx.Done()
}

// Stop is a no-op since each Connect call manages its own connection lifecycle.
func (c *WSConnector) Stop() {}
